File: stack_queue/st_in_st/st_in_st.cpp
#include <iostream>
#include <vector>
#include <stdexcept>

using namespace std;

/**********************************************
 * CATALOG OF FUNCTIONS AND CLASSES
 **********************************************
 * 
 * CLASS: Stack<T>
 * 
 * 1. `void push(const T& value)`
 *    - Adds an element `value` to the top of the stack.
 *    - Dynamically resizes the stack if needed.
 * 
 * 2. `void pop()`
 *    - Removes the top element from the stack.
 *    - Throws a `runtime_error` if the stack is empty.
 * 
 * 3. `T& top()`
 *    - Returns a reference to the top element of the stack.
 *    - Throws a `runtime_error` if the stack is empty.
 * 
 * 4. `bool isEmpty() const`
 *    - Checks if the stack is empty.
 *    - Returns `true` if empty, `false` otherwise.
 * 
 * 5. `int size() const`
 *    - Returns the number of elements in the stack.
 * 
 * 6. `void print() const`
 *    - Prints the elements of the stack from bottom to top.
 * 
 * 7. `const vector<T>& getData() const`
 *    - Returns a reference to the underlying data vector of the stack.
 *    - Useful for operations like merging stacks.
 * 
 **********************************************
 * FUNCTION: mergeInnerStacks
 * 
 * 1. `void mergeInnerStacks(Stack<Stack<T>>& stackOfStacks, int i, int j, int k)`
 *    - Merges two inner stacks at indices `i` and `j` into a new stack.
 *    - Inserts the merged stack at position `k` in the stack of stacks.
 *    - Dynamically adjusts the stack of stacks to accommodate the merged stack.
 *    - **Parameters**:
 *       - `stackOfStacks`: The stack of stacks containing the inner stacks.
 *       - `i`: The index of the first inner stack to merge.
 *       - `j`: The index of the second inner stack to merge.
 *       - `k`: The position to insert the merged stack.
 *    - **Error Handling**:
 *       - Throws a `runtime_error` if any of the indices are invalid.
 **********************************************/

// Template class for a Stack
template <typename T>
class Stack {
private:
    vector<T> data; // Vector to store stack elements

public:
    // Push an element onto the stack
    void push(const T& value) {
        data.push_back(value);
    }

    // Pop the top element from the stack
    void pop() {
        if (isEmpty())
            throw runtime_error("Stack underflow");
        data.pop_back();
    }

    // Get the top element of the stack
    T& top() {
        if (isEmpty())
            throw runtime_error("Stack is empty");
        return data.back();
    }

    // Check if the stack is empty
    bool isEmpty() const {
        return data.empty();
    }

    // Get the size of the stack
    int size() const {
        return data.size();
    }

    // Print the stack contents
    void print() const {
        for (const auto& elem : data)
            cout << elem << " ";
        cout << endl;
    }

    // Access the underlying data (needed for merging)
    const vector<T>& getData() const {
        return data;
    }
};

// Function to merge two inner stacks and insert the merged stack into a stack of stacks
template <typename T>
void mergeInnerStacks(Stack<Stack<T>>& stackOfStacks, int i, int j, int k) {
    if (i >= stackOfStacks.size() || j >= stackOfStacks.size() || k > stackOfStacks.size() || i == j)
        throw runtime_error("Invalid indices");

    Stack<T> mergedStack;

    Stack<T> stack1 = stackOfStacks.getData()[i];
    Stack<T> stack2 = stackOfStacks.getData()[j];

    for (const T& value : stack1.getData())
        mergedStack.push(value);

    for (const T& value : stack2.getData())
        mergedStack.push(value);

    auto& stacks = const_cast<vector<Stack<T>>&>(stackOfStacks.getData());
    if (i > j) swap(i, j); // Ensure proper order for deletion
    stacks.erase(stacks.begin() + j);
    stacks.erase(stacks.begin() + i);
    stacks.insert(stacks.begin() + k, mergedStack);
}

// Main function for testing
int main() {
    cout << "=== Testing Inner Stack Functions ===" << endl;

    // Create an inner stack
    Stack<int> innerStack;
    cout << "Inner stack created." << endl;

    // Push elements
    innerStack.push(10);
    innerStack.push(20);
    innerStack.push(30);
    cout << "Pushed elements onto the inner stack: 10, 20, 30." << endl;

    // Peek top element
    cout << "Top element of inner stack: " << innerStack.top() << endl;

    // Pop elements
    cout << "Popped element: " << innerStack.top() << endl;
    innerStack.pop();
    cout << "Popped element: " << innerStack.top() << endl;
    innerStack.pop();

    // Check if stack is empty
    cout << "Is inner stack empty? " << (innerStack.isEmpty() ? "Yes" : "No") << endl;

    cout << "\n=== Testing Outer Stack (Stack of Stacks) ===" << endl;

    // Create an outer stack
    Stack<Stack<int>> outerStack;
    cout << "Outer stack created." << endl;

    // Create inner stacks
    Stack<int> innerStack1, innerStack2;
    innerStack1.push(1);
    innerStack1.push(2);
    innerStack2.push(3);
    innerStack2.push(4);

    // Push inner stacks onto the outer stack
    outerStack.push(innerStack1);
    outerStack.push(innerStack2);
    cout << "Inner stacks pushed onto the outer stack." << endl;

    // Print outer stack
    cout << "Outer stack contents:" << endl;
    for (int i = 0; i < outerStack.size(); i++) {
        cout << "Stack " << i << ": ";
        outerStack.getData()[i].print();
    }

    // Peek the top inner stack
    cout << "Top inner stack (peek): ";
    outerStack.top().print();

    // Pop the top inner stack
    Stack<int> poppedStack = outerStack.top();
    outerStack.pop();
    cout << "Popped top inner stack. Contents of popped stack: ";
    poppedStack.print();

    // Check if the outer stack is empty
    cout << "Is outer stack empty? " << (outerStack.isEmpty() ? "Yes" : "No") << endl;

    cout << "\n=== Testing Merge Inner Stacks ===" << endl;

    // Push innerStack2 back and create another stack
    outerStack.push(innerStack2);
    Stack<int> innerStack3;
    innerStack3.push(5);
    innerStack3.push(6);
    outerStack.push(innerStack3);

    // Print before merging
    cout << "Outer stack before merging:" << endl;
    for (int i = 0; i < outerStack.size(); i++) {
        cout << "Stack " << i << ": ";
        outerStack.getData()[i].print();
    }

    // Merge two inner stacks
    mergeInnerStacks(outerStack, 0, 1, 1);

    // Print after merging
    cout << "Outer stack after merging inner stacks 0 and 1 into position 1:" << endl;
    for (int i = 0; i < outerStack.size(); i++) {
        cout << "Stack " << i << ": ";
        outerStack.getData()[i].print();
    }

    return 0;
}

File: stack_queue/st_in_st/st_in_st.py
"""
**********************************************
* CATALOG OF FUNCTIONS AND CLASSES
**********************************************

CLASS: Stack
1. __init__(self):
   - Initializes an empty stack.
   - Creates an internal list to hold stack elements.

2. push(self, value):
   - Adds an element `value` to the top of the stack.

3. pop(self):
   - Removes and returns the top element of the stack.
   - Raises `IndexError` if the stack is empty.

4. top(self):
   - Returns the top element of the stack without removing it.
   - Raises `IndexError` if the stack is empty.

5. is_empty(self):
   - Checks if the stack is empty.
   - Returns `True` if the stack is empty, otherwise `False`.

6. size(self):
   - Returns the number of elements in the stack.

7. print_stack(self):
   - Prints the elements of the stack from bottom to top.

8. get_data(self):
   - Returns the underlying data list of the stack.
   - Useful for operations such as merging.

**********************************************

FUNCTION: merge_inner_stacks
1. merge_inner_stacks(stack_of_stacks, i, j, k):
   - Merges two inner stacks at indices `i` and `j` from `stack_of_stacks`.
   - Deletes the original stacks at indices `i` and `j`.
   - Inserts the merged stack into `stack_of_stacks` at position `k`.

   **Parameters**:
      - `stack_of_stacks`: A `Stack` of `Stack` objects.
      - `i`: Index of the first inner stack to merge.
      - `j`: Index of the second inner stack to merge.
      - `k`: Position to insert the merged stack.

   **Error Handling**:
      - Raises `IndexError` if any of the indices are invalid.
**********************************************
"""

class Stack:
    def __init__(self):
        """
        Initializes an empty stack.
        """
        self.data = []

    def push(self, value):
        """
        Push an element onto the stack.

        Parameters:
            value: The value to be added to the stack.
        """
        self.data.append(value)

    def pop(self):
        """
        Pop the top element from the stack.

        Returns:
            The popped element.

        Raises:
            IndexError: If the stack is empty.
        """
        if self.is_empty():
            raise IndexError("Stack underflow")
        return self.data.pop()

    def top(self):
        """
        Get the top element of the stack without removing it.

        Returns:
            The top element of the stack.

        Raises:
            IndexError: If the stack is empty.
        """
        if self.is_empty():
            raise IndexError("Stack is empty")
        return self.data[-1]

    def is_empty(self):
        """
        Check if the stack is empty.

        Returns:
            True if the stack is empty, False otherwise.
        """
        return len(self.data) == 0

    def size(self):
        """
        Get the size of the stack.

        Returns:
            The number of elements in the stack.
        """
        return len(self.data)

    def print_stack(self):
        """
        Print the stack contents from bottom to top.
        """
        print(" ".join(map(str, self.data)))

    def get_data(self):
        """
        Access the underlying data of the stack.

        Returns:
            A list representing the stack's contents.
        """
        return self.data


def merge_inner_stacks(stack_of_stacks, i, j, k):
    """
    Merge two inner stacks and insert the merged stack into the stack of stacks.

    Parameters:
        stack_of_stacks (Stack): The stack of stacks containing the inner stacks.
        i (int): The index of the first inner stack to merge.
        j (int): The index of the second inner stack to merge.
        k (int): The position to insert the merged stack.

    Raises:
        IndexError: If any of the indices are invalid.
    """
    if i >= stack_of_stacks.size() or j >= stack_of_stacks.size() or k > stack_of_stacks.size() or i == j:
        raise IndexError("Invalid indices")

    # Ensure i is smaller than j for proper deletion
    if i > j:
        i, j = j, i

    # Create a new stack to hold the merged data
    merged_stack = Stack()

    # Retrieve stacks i and j
    stack1 = stack_of_stacks.get_data().pop(i)
    stack2 = stack_of_stacks.get_data().pop(j - 1)

    # Push elements from the first stack
    for value in stack1.get_data():
        merged_stack.push(value)

    # Push elements from the second stack
    for value in stack2.get_data():
        merged_stack.push(value)

    # Insert the merged stack at position k
    stack_of_stacks.get_data().insert(k, merged_stack)


if __name__ == "__main__":
    # Create a stack of stacks
    stack_of_stacks = Stack()

    # Create the first inner stack
    stack1 = Stack()
    stack1.push(1)
    stack1.push(2)

    # Create the second inner stack
    stack2 = Stack()
    stack2.push(3)
    stack2.push(4)
    
    # Create the second inner stack
    stack3 = Stack()
    stack3.push(5)
    stack3.push(6)

    # Push the inner stacks onto the stack of stacks
    stack_of_stacks.push(stack1)
    stack_of_stacks.push(stack2)
    stack_of_stacks.push(stack3)

    # Print the stack of stacks before merging
    print("Before merging:")
    for idx, stack in enumerate(stack_of_stacks.get_data()):
        print(f"Stack {idx}: ", end="")
        stack.print_stack()

    # Merge stack1 and stack2 into a new stack at position 1
    merge_inner_stacks(stack_of_stacks, 0, 1, 1)

    # Print the stack of stacks after merging
    print("\nAfter merging:")
    for idx, stack in enumerate(stack_of_stacks.get_data()):
        print(f"Stack {idx}: ", end="")
        stack.print_stack()